docs/decisions/0031-automatic-tool-calling-controls.md

```diff
@@ -0,0 +1,198 @@
+---
+# These are optional elements. Feel free to remove any of them.
+status: proposed
+contact: gitri-ms
+date: 2024-01-19
+deciders: gitri-ms, stephentoub, alliscode, markwallace-microsoft
+consulted: dmytrostruk
+informed:
+---
+
+# Automatic Tool Calling Controls
+
+## Context and Problem Statement
+
+[Task 4300](https://github.com/microsoft/semantic-kernel/issues/4300) is underway to use automatic tool calling in the `FunctionCallingStepwisePlanner`. This will allow for more robust and consistent handling of tool calls and minimize duplication of code.
+
+Currently, as part of the planner configuration, the caller can set the `MaxIterations` field to limit the number of steps (and therefore, calls to the model) utilized by the planner. This enables the caller to control things such as cost, time, etc., spent by the planner.
+
+The current implementation of automatic tool calling allows for multiple consecutive tool calls to be requested/invoked by the model for a single request. However, the limit on how many consecutive tool calls can be performed per request is not exposed to the caller. Therefore, the caller cannot control or even measure after the fact how many model roundtrips occur for a single request. This control is necessary for the planner to accurately enforce the `MaxIterations` limit.
+
+## Decision Drivers
+
+1. Developers should be able to control how many tool calls (and subsequent model calls) are performed for a single request
```

Comment:
We should think about the various knobs someone will need, beyond just what the planner needs right now. Even if we don't expose all the knobs, we should implement this in a way that makes it easy for us to expose additional stuff in the future. For example, https://github.com/microsoft/semantic-kernel/issues/4496#issuecomment-1879175887 effectively wants to be able to examine which function is being invoked, and decide to allow the invocation but then stop further processing. That's more than just MaxIterations, but likely some kind of callback scheme where the callback is presented with a bag of information, like how many iterations have occurred, what function is being called, what arguments are being passed to it, etc., and the callback can then decide to allow the function to invoke, or allow it to invoke but end the processing with its results, or not let it be invoked at all, etc.  We might even want to allow it to do the invocation, so that it could have more control over how it's invoked, or even what's invoked, or what happens if it fails.

Reply:
These requirements sound similar to what we already have in filtering functionality. Can we leverage function filters to cover this scenario? We can extend the context of filter with more information if needed, like iteration number, etc, so it would be possible to make a decision about function execution in the scope of automatic tool calling.